### ISSUE TITLE
Refactor the Event API into a more promises like syntax

### ISSUE THREAD

**Issue by Inkdpixels** [reduct/component#1]
The goal is to have a more flexible system for subscribing and publishing of events. The current event dispatcher falls short in some cases and the overall code difference to a full PubSub system isn't that much larger.

**Comment by Inkdpixels** [comment]
As discussed yesterday and noted [here](https://github.com/reduct/discussions/wiki/%5B28.07.2015%5D-Protocol#component) - We should strive into a direction where the Event API should have a more promises based syntax. 

This could be achieved with generators/promises.

**Comment by Inkdpixels** [comment]
Just thought about this. 

If we implement this with the full functionality and spec conformity of generators or promises, we require that all developers who want to use this package will include a polyfill for the used es6 feature. This is as of now a big decision to make, and I am unsure how to proceed here.

Either we go the route with the requirement of polyfills - Which of course also has it's benefits. 
Or we implement a subset or just the syntax of promises.

What are your thoughts about this @akoenig @grebaldi?

A proposal on how the API could look/work.

``` javascript
class MyComponent extends ReductComponent {
    constructor() {
        this.on('myActionType').then((args...) => {
            // Do something.
        });
    }
    myMethod() {
        this.dispatch('myActionType', (resolve, reject) => {
            // Do something and resolve / reject the promise.
        });
    }
}
```

The second argument of the dispatch method should be optional, and not strictly bound to a function. If a function was given, we should await the execution of the resolve or reject callbacks. And if something else was provided as a second argument, the Event will automatically fire and will use the second argument as the arguments for the resolver.

Is this reasonable?

**Comment by grebaldi** [comment]
I don't think it's necessary to create actual promises for the events themselves -> except there's any chance an event might need some "reject"-semantics, which I don't see atm.

Event emitting should work like one's used to: just a simple method call with an event name and payload. For the event handling side, we should aim for a pretty much Promises/A+ compliant API, so that further logic could be organized in a promise-typical `then`-chain. (At some point the creation of an actual Promise would be helpful, though)

Maybe it would be a solution to offer the user injection for the promise implementation of his/her choice via factoryOpts -> and then fall back to native Promises if no explicit override is defined.

Besides that, `on` might need a renaming. `listen` or `waitFor` make more sense in a Promise context.

**Comment by akoenig** [comment]
Hm, I'm also against forcing the user to provide a promise polyfill of her choice. This would result in "Hey, you have to provide this, so that this library can do it's job." Which would result in a lower framework adoption in the end: "WTF? I have to resolve the dependencies on my own?"

I think we shouldn't provide an API for manually injecting the respective promise implementation as well. A simple: `let Promise = global.Promise || ourPromiseImpl;` would do the job as every established polyfill is available via the global object anyway.

Long story short: I'm in favor of implementing an own lightweight Promise/A+ compliant API layer which in fact mimics the actual API. An important aspect is to handle immediate available `resolvements` via `setImmediate` or a proper `process.nextTick` implementation for not blocking the event loop and therefore releasing the [Zalgo](http://blog.izs.me/post/59142742143/designing-apis-for-asynchrony).

**Comment by Inkdpixels** [comment]
As discussed yesterday, this wouldn't solve the problem. Afaik the problem itself is not existing anymore.

**Comment by akoenig** [comment]
What do you mean by "this wouldn't solve the problem"? - As far as I remember, we discussed that we will stick with the current implementation.

My suggestion was regarding solving problems which comes with implementing an own thin Promise layer.

**Comment by Inkdpixels** [comment]
Yeah totally, this was more aimed on the general topic - Promises wouldn't solve the so thought issue, which isn't an issue as we've discussed. :)
